<a href="https://colab.research.google.com/github/datakind/hxl-metadata-prediction/blob/main/openai-hxl-prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

This notebook provides examples of predicting HXL tags and attributes by (i) Fine-tuning GPT-4o; (ii) Direct prompting using the HXL standard.




# Key findings

Fine-tuning an LLM to predict HXL tags is low-cost, but it constrained by the training data. Many datasets tagged with HXL has used tags and attributes which are not part of the HXL standard, and though there is good representation of tags and attributes related to location and dates, which are indeed very useful, beyond this coverage is poor. These challenges in the training data limit the scenarios in which the LLM fine-tuning technique can be applied.

However, in recent months there have been significant advances in LLMs, for example with the release of GPT-4o. Using this model, prompted with the full HXL standard definitions, it was shown that performance is better than fine-tuning and in some cases better than the humans who tagged the training data.



# Setup

See [README](README.md) for instructions.

In [ ]:
!pip install pandas==2.2.2
!pip install openai==1.35.3
!pip install python-dotenv==1.0.1

In [100]:
import openai
import os
import time
import openai
from openai import OpenAI
import pandas as pd
import json
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score
import sys
import requests
import pprint

from dotenv import load_dotenv
load_dotenv()

if os.getenv("OPENAI_API_KEY") is None:
  from google.colab import userdata
  OPENAI_API_KEY =  userdata.get('OPENAI_API_KEY')
else:
  OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")

client = OpenAI(
    api_key=OPENAI_API_KEY
)

# If using Colab, this is where Google drive gets mounted. Otherwise leave blank
GOOGLE_BASE_DIR = "/content/drive/MyDrive/Colab"

# This is the HXL schema sheet, search HDX to get this link
HXL_SCHEMA_RESOURCE_URL = "https://docs.google.com/spreadsheets/d/1En9FlmM8PrbTWgl3UHPF_MXnJ6ziVZFhBbojSJzBdLI/export?format=xlsx"

# Where to save local data files
LOCAL_DATA_DIR = f"{GOOGLE_BASE_DIR}/hxl-metadata-prediction/data/"

# As generated by generate-test-train-data.ipynb
TRAINING_FILE = f"{LOCAL_DATA_DIR}/hxl_chat_prompts_train.jsonl"
TEST_FILE = f"{LOCAL_DATA_DIR}/hxl_chat_prompts_test.jsonl"

# Base model to fine-tune
MODEL = "gpt-4o-mini-2024-07-18"

pd.set_option('display.max_colwidth', 900)
pd.set_option('display.max_rows', 50)


In [101]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Download HXL Schema

In [102]:
local_data_file = LOCAL_DATA_DIR + "/hxl-core-schema.xlsx"

response = requests.get(HXL_SCHEMA_RESOURCE_URL)
with open(local_data_file, 'wb') as f:
    f.write(response.content)

df= pd.read_excel(local_data_file, sheet_name='Core hashtags')
hashtags_list = df['Hashtag'][1:].tolist()

df= pd.read_excel(local_data_file, sheet_name='Core attributes')
attributes_list = df['Attribute'][1:].tolist()

# Remove rows with disallowed tags or attributes
APPROVED_HXL_SCHEMA = hashtags_list + attributes_list

print("Approved HXL schema ...")
print(APPROVED_HXL_SCHEMA)

Approved HXL schema ...
['#access', '#activity', '#adm1', '#adm2', '#adm3', '#adm4', '#adm5', '#affected', '#beneficiary', '#capacity', '#cause', '#channel', '#contact', '#country', '#crisis', '#currency', '#date', '#delivery', '#description', '#event', '#frequency', '#geo', '#group', '#impact', '#indicator', '#inneed', '#item', '#loc', '#meta', '#modality', '#need', '#operations', '#org', '#output', '#population', '#reached', '#region', '#respondee', '#sector', '#service', '#severity', '#status', '#subsector', '#targeted', '#value', '+abducted', '+acronym', '+activity', '+adolescents', '+adults', '+approved', '+ar', '+bounds', '+budget', '+canceled', '+children', '+cluster', '+code', '+converted', '+coord', '+dest', '+displaced', '+elderly', '+elevation', '+email', '+en', '+end', '+es', '+f', '+fa', '+fr', '+funder', '+hh', '+i', '+id', '+idps', '+impl', '+incamp', '+ind', '+infants', '+infected', '+injured', '+killed', '+label', '+lat', '+lon', '+m', '+ms', '+name', '+noncamp', '+num

# Analysis

## Check test/train

Let's do a sanity check to ensure the test set doesn't include data from orgnaizations in the traning set.

## Fine-tune

In [ ]:
def fine_tune_model(train_file, model_name="gpt-4o-mini"):
    """
    Fine-tune an OpenAI model using training data.

    Args:
        prompt_file (str): The file containing the prompts to use for fine-tuning.
        model_name (str): The name of the model to fine-tune. Default is "davinci-002".

    Returns:
        str: The ID of the fine-tuned model.
    """

    # Create a version of the train_file jsonl which only has "messages"
    train_file_short = train_file.replace(".jsonl", "_short.jsonl")
    with open(train_file) as f:
        prompts = [json.loads(line) for line in f]
        prompts = [p["messages"] for p in prompts]
        with open(train_file_short, "w") as f:
            for p in prompts:
              row = {}
              row["messages"] = p
              f.write(json.dumps(row) + "\n")

    # Upload file to OpenAI for fine-tuning
    file = client.files.create(
        file=open(train_file_short, "rb"),
        purpose="fine-tune"
    )
    file_id = file.id
    print(f"Uploaded training file with ID: {file_id}")

    # Start the fine-tuning job
    ft = client.fine_tuning.jobs.create(
        training_file=file_id,
        model=model_name
    )
    ft_id = ft.id
    print(f"Fine-tuning job started with ID: {ft_id}")

    # Monitor the status of the fine-tuning job
    ft_result = client.fine_tuning.jobs.retrieve(ft_id)
    while ft_result.status != 'succeeded':
        print(f"Current status: {ft_result.status}")
        time.sleep(120)  # Wait for 60 seconds before checking again
        ft_result = client.fine_tuning.jobs.retrieve(ft_id)
        if 'failed' in ft_result.status.lower():
            sys.exit()

    print(f"Fine-tuning job {ft_id} succeeded!")

    # Retrieve the fine-tuned model
    fine_tuned_model = ft_result.fine_tuned_model
    print(f"Fine-tuned model: {fine_tuned_model}")

    return fine_tuned_model

In [103]:
def read_prompts_file(filename):
  results = []
  with open(filename) as f:
    prompts = [json.loads(line) for line in f]
    for p in prompts:
      p["prompt"] = p["messages"][0:2]
      p["expected"] = p["messages"][-1]["content"]
      results.append(p)
    results = pd.DataFrame(results)

    print(f"\nFound {len(results)} prompts")
    print(f"\nData providers {results['Data provider'].unique()}")

    results['tag'] = results['expected'].apply(lambda x: x.split('+')[0])
    tag_counts = results['tag'].value_counts()
    print("\n",tag_counts)

  return results

print("\n\n======= TRAIN =======")
X_train = read_prompts_file(TRAINING_FILE)
print("\n\n======= TRAIN =======")
X_test = read_prompts_file(TEST_FILE)

# Print data providers in X_test that are in X_train
common_providers = list(set(X_train["Data provider"]).intersection(set(X_test["Data provider"])))
if len(common_providers) == 0:
  print("No common Data providers")
else:
  print(f"Common providers: {common_providers} found in both Train and test sets!!!")
  sys.exit()

# Print any hashes in test which are in train
common_hashes = list(set(X_train["expected"]).intersection(set(X_test["expected"])))
if len(common_hashes) == 0:
  print("No common hashes")

display(X_train)



======= TRAIN =======

Found 2919 prompts

Data providers ['international-organization-for-migration'
 'eth-zurich-weather-and-climate-risks' 'ifrc' 'ocha-fts' 'cerf' 'awsd'
 'insecurity-insight' 'ocha-sudan' 'ocha-niger' 'wfp' 'ocha-car' 'cred'
 'fao' 'water-point-data-exchange' 'ipc' 'interaction' 'ocha-somalia'
 'hdx' 'ocha-yemen' 'ocha-afghanistan' 'ourairports' 'hxl'
 'world-bank-group' 'unrwa-for-palestine-refugees-in-the-near-east'
 'ocha-fiss' 'ocha-ukraine' 'unhcr' 'ocha-ethiopia' 'ocha-haiti'
 'ocha-colombia' 'ocha-chad' 'ocha-nigeria' 'ocha-myanmar'
 'ocha-south-sudan' 'ocha-mali' 'ocha-dr-congo'
 'blavatnik-school-of-government-university-of-oxford' 'ocha-burkina'
 'un-ocha' 'ocha-ds' 'reliefweb' 'ocha-rosc' 'ocha-cameroon' 'unicef-rdc'
 'ocha-rosea' 'ocha-rolac' 'ocha-burundi' 'world-health-organization'
 'jcc' 'international-displacement-monitoring-centre-idmc' 'ocha-iraq'
 'ocha-opt' 'qcri' 'health-cluster' 'ocha-mozambique-hat' 'unicef-data'
 'unesco' 'ocha-libya' 'oc

messages  \
0     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx'; dataset_description='The dataset from the file "DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx" contains information on the total number of internally displaced persons (IDPs) and returnees in the Democratic Republic of Congo. The data includes the total number of IDP households, IDP individuals, male and female IDPs, and returnees. Specifically, there are 319,283 IDP households, 1,548,732 IDP individuals, with 646,805 males and 901,927 females, and 587,705 returnees.'; column_name:'Total IDP HH'; examples: [319283]'}, {'rol...   
1     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx'; dataset_description='The dataset from the file "DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx" contains information on the total number of internally displaced persons (IDPs) and returnees in the Democratic Republic of Congo. The data includes the total number of IDP households, IDP individuals, male and female IDPs, and returnees. Specifically, there are 319,283 IDP households, 1,548,732 IDP individuals, with 646,805 males and 901,927 females, and 587,705 returnees.'; column_name:'Total Returnees'; examples: [587705]'}, {'...   
2     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The dataset contains earthquake data for various administrative regions in Afghanistan, including country name, admin1 name, latitude, longitude, aggregation type, indicator name, and indicator value. The data includes maximum earthquake values recorded in different regions, with corresponding latitude and longitude coordinates. The dataset provides insights into the seismic activity in different administrative areas of Afghanistan.'; column_name:'country_name'; examples: ['Afghanistan', 'Afghanistan', 'Afghanistan', 'Afghanistan', 'Afghanistan', 'Af...   
3     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The dataset contains earthquake data for various administrative regions in Afghanistan, including country name, admin1 name, latitude, longitude, aggregation type, indicator name, and indicator value. The data includes maximum earthquake values recorded in different regions, with corresponding latitude and longitude coordinates. The dataset provides insights into the seismic activity in different administrative areas of Afghanistan.'; column_name:'latitude'; examples: ['34.5527', '34.9568', '34.9619', '34.3033', '34.0121', '34.2743', '34.7693', '35.4...   
4     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The da

In [ ]:
model = fine_tune_model(TRAINING_FILE, model_name=MODEL)

Uploaded training file with ID: file-Uzsels8NXSGHecUo2672vxG5
Fine-tuning job started with ID: ftjob-TxNTfk1vI83R7dNor0rqlVkU
Current status: validating_files
Current status: validating_files
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Current status: running
Fine-tuning job ftjob-TxNTfk1vI83R7dNor0rqlVkU succeeded!
Fine-tuned model: ft:gpt-4o-mini-2024-07-18:datakind::9p1xodpF


In [ ]:
#model = "ft:gpt-4o-mini-2024-07-18:datakind::9oJXzcfa" # No data summaries
model = "ft:gpt-4o-mini-2024-07-18:datakind::9p1xodpF # With data description"
print(f"Fine-tuned model: {model}")

Fine-tuned model: ft:gpt-4o-mini-2024-07-18:datakind::9p1xodpF # With data description


## Prediction Test

In [ ]:
def make_chat_predictions(prompts, model, temperature=0.1, max_tokens=13):
  results = []
  for p in prompts:
    actual = p["messages"][-1]["content"]
    p["messages"] = p["messages"][0:2]
    completion = client.chat.completions.create(
      model=model,
      messages=p["messages"],
      temperature=temperature,
      max_tokens=max_tokens
    )
    predicted = completion.choices[0].message.content
    predicted = filter_for_schema(predicted)

    res = {
        "prompt": p["messages"],
        "actual": actual,
        "predicted": predicted
    }

    print(f"Predicted: {predicted}; Actual: {actual}")

    results.append(res)

  results = pd.DataFrame(results)

  return results

def filter_for_schema(text):
    #print(f"Tokens before: {text}")
    if " " in text:
        text = text.replace(" ","")

    tokens_raw = text.split("+")
    tokens = [tokens_raw[0]]
    for t in tokens_raw[1:]:
        tokens.append(f"+{t}")

    filtered = []
    for t in tokens:
        if t in APPROVED_HXL_SCHEMA:
            if t not in filtered:
                filtered.append(t)
    filtered = "".join(filtered)

    if len(filtered) > 0 and filtered[0] != '#':
        filtered = ""

    # Add spaces back in
    # filtered = filtered.replace("+", " +")

    #print(f"        After: {filtered}")
    return filtered

def output_prediction_metrics(results, prediction_field="predicted", actual_field="actual"):
    """
    Prints out model performance report.

    Parameters
    ----------
    results : dataframe
        Dataframe of results
    prediction_field : str
        Field name of element with prediction. Handy for comparing raw and post-processed predictions.
    actual_field: str
        Field name of the actual result for comparison with prediction
    """
    y_test = []
    y_pred = []
    y_justtag_test = []
    y_justtag_pred = []
    for index, r in results.iterrows():
        if actual_field not in r and predicted_field not in r:
            print("Provided results do not contain expected values.")
            sys.exit()
        y_pred.append(r[prediction_field])
        y_test.append(r[actual_field])
        actual_tag = r[actual_field].split("+")[0]
        predicted_tag = r[prediction_field].split("+")[0]
        y_justtag_test.append(actual_tag)
        y_justtag_pred.append(predicted_tag)

    print(f"LLM results for {prediction_field}, {len(results)} predictions ...")
    print("\nJust HXL tags ...\n")
    print(f"Accuracy: {round(accuracy_score(y_justtag_test, y_justtag_pred),2)}")
    print(
        f"Precision: {round(precision_score(y_justtag_test, y_justtag_pred, average='weighted', zero_division=0),2)}"
    )
    print(
        f"Recall: {round(recall_score(y_justtag_test, y_justtag_pred, average='weighted', zero_division=0),2)}"
    )
    print(
        f"F1: {round(f1_score(y_justtag_test, y_justtag_pred, average='weighted', zero_division=0),2)}"
    )

    print(f"\nTags and attributes with {prediction_field} ...\n")
    print(f"Accuracy: {round(accuracy_score(y_test, y_pred),2)}")
    print(
        f"Precision: {round(precision_score(y_test, y_pred, average='weighted', zero_division=0),2)}"
    )
    print(
        f"Recall: {round(recall_score(y_test, y_pred, average='weighted', zero_division=0),2)}"
    )
    print(
        f"F1: {round(f1_score(y_test, y_pred, average='weighted', zero_division=0),2)}"
    )

    return

In [ ]:
with open(TEST_FILE) as f:
    X_test = [json.loads(line) for line in f]

# Subsample
#size = 10
#X_test = X_test[-size:]

results = make_chat_predictions(X_test, model)

results.to_excel(f"{LOCAL_DATA_DIR}/hxl-metadata-prediction-results.xlsx", index=False)

display(results)

output_prediction_metrics(results)

print("Done")

Predicted: #country+code; Actual: #country+code
Predicted: #adm1+name; Actual: #loc+name
Predicted: #indicator+code; Actual: #meta+id
Predicted: #indicator+name; Actual: #indicator+name
Predicted: #country+name; Actual: #country+name
Predicted: #indicator+code; Actual: #indicator+code
Predicted: #indicator+code; Actual: #indicator+code+label
Predicted: #country+code; Actual: #country+code
Predicted: #meta+id; Actual: #meta+id
Predicted: #indicator+name; Actual: #indicator+name
Predicted: #country+name; Actual: #country+name
Predicted: #indicator+code; Actual: #indicator+code
Predicted: #indicator+code; Actual: #indicator+code+label
Predicted: #indicator+name; Actual: #indicator+label
Predicted: #country+code; Actual: #country+code
Predicted: #country+name; Actual: #country+name
Predicted: #indicator+code; Actual: #indicator+id
Predicted: #indicator+name; Actual: #indicator+name
Predicted: #country+code; Actual: #country+code
Predicted: #country+name; Actual: #country+name
Predicted: #i

prompt  \
0    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
1    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
2    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
3    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
4    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_descript

Done


### Prediction analysis


Next we will look at cases where the prediction failed to see patterns to address.

In [ ]:
results["match"] = results['predicted'] == results['actual']
display(results[results["match"]==False])

prompt  \
1    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
6    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
12   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provi...   
13   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provi...   
16   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Human Development Indicators for Zimbabwe.csv'; dataset_description=

#### Scenario 1 - Predicting 'adm1' when it should region

Better prompting could fix this, to inform that region != adm1.

In [ ]:
# Find rows where match is False and predicted contains admin1
breaks = results[results["match"]==False]
print(breaks.shape)
scenario1 = breaks[(breaks["match"]==False) & (breaks["actual"].str.contains("region")) & (breaks["predicted"].str.contains("adm1"))]
display(scenario1)
print(scenario1.shape)

(157, 4)


prompt  \
26                                                                                                                                                                                                                                                                                                                                                           [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Burundi- Muyinga, Cankuzo, Makamba, Ruyigi, Rutana, Rumonge: Operational Presence.xlsx'; dataset_description='None'; column_name:'Region'; examples: ['Muyinga', 'Muyinga', 'Muyinga', 'Muyinga', 'Cankuzo', 'Rumonge', 'Ruyigi', 'Ruyigi', 'Makamba', 'Makamba', 'Rutana']'}]   
60   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/NGA_Subnational_Covid19_HXL_HERA.csv'; dataset_description='The dataset contains information on COVID-19 cases in different regions of Nigeria, with columns including ID, date, ISO code, country name, region name, number of confirmed cases, deaths, recoveries, and gender-specific case counts. The data is sourced from the Nigeria Centre for Disease Control and covers various regions such as Abia, Adamawa, Akwa Ibom, Anambra, Bauchi, Bayelsa, Benue, Borno, and Cross River. The dataset appears to be structured with semicolons as delimiters and includes metadata headers at the beginning.'; column_name:'REGIO...   
66   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/BFA_Subnational_Covid19_HXL_HERA.csv'; dataset_description='The dataset contains information on COVID-19 cases in Burkina Faso at a subnational level, with columns including ID, DATE, ISO_3, PAYS, ID_PAYS, REGION, ID_REGION, CONTAMINES, DECES, GUERIS, CONTAMINES_FEMME, CONTAMINES_HOMME, CONTAMINES_GENRE_NON_SPECIFIE, NOUVEAUX_INDIVIDUS_VACCINES_1DOSE, TOTAL_INDIVIDUS_VACCINES_1DOSE, NOUVEAUX_AGENTS_SANTE_VACCINES_1DOSE, TOTAL_AGENTS_SANTE_VACCINES_1DOSE, NOUVEAUX_INDIVIDUS_VACCINES_2DOSES, TOTAL_INDIVIDUS_VACCINES_2DOSES, NOUVEAUX_AGENTS_SANTE_VACCINES_2DOSES, TOTAL_AGENTS_SANTE_VACCINES_2DOSES, and SOUR...   
72   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/MRT_Subnational_Covid19_HXL_HERA.csv'; dataset_description='The dataset contains information on COVID-19 cases in Mauritania at a subnational level, with columns including ID, date, ISO code, country name, region, number of confirmed cases, deaths, recoveries, and gender breakdown of cases. The data is sourced from the Ministry of Health. The dataset provides details for different regions within Mauritania, with each row representing a specific region and its corresponding COVID-19 statistics on a particular date.'; column_name:'REGION'; examples: ['Adrar', 'Assaba', 'Brakna', 'Dakhlet Nouadhibou', 'Gorg...   
78   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/MLI_Subnational_Covid19_HXL_HERA.csv'; dataset_description='The data

(17, 4)


#### Scenario 2 - When labeled data says admin1 but really it is country (admin 0)

In [ ]:
breaks = breaks[~breaks.index.isin(scenario1.index)]
print(breaks.shape)

scenario2 = breaks[(breaks["predicted"].str.contains("country")) & (breaks["actual"].str.contains("adm1"))]
display(scenario2)
print(scenario2.shape)


(140, 4)


prompt  \
25                                                                                                                                                                                                                                                                                                                                                       [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Burundi- Muyinga, Cankuzo, Makamba, Ruyigi, Rutana, Rumonge: Operational Presence.xlsx'; dataset_description='None'; column_name:'Country'; examples: ['Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi', 'Burundi']'}]   
38                                                                                                                                                                                                                                                                                                                                                                     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/FieldsData_3W_PHL_Aklan.xlsx'; dataset_description='None'; column_name:'Country'; examples: ['Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines', 'Philippines']'}]   
226  [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Kerela.xlsx'; dataset_description='The dataset from the file "Kerela.xlsx" contains information about various organizations in South India, specifically in Wayanad, India. The data includes details such as the timestamp, source, name of the organization, type of organization, region, country, province, and sector. The organizations mentioned are involved in areas such as education, health & nutrition, livelihoods, environment protection, and food security. The dataset seems to capture activities and initiatives undertaken by these organizations in the specified region.'; column_name:'Country'; examples: ...   
327  [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Burundi Cankuzo_FD.xlsx'; dataset_description='The dataset from the file "Burundi Cankuzo_FD.xlsx" contains information on various organizations operating in the Cankuzo region of Burundi. The data includes details such as the timestamp, data source, organization name, region, country, province, and sector of each organization. Organizations mentioned in the dataset are involved in sectors like Food Security, Protection, SGBV, Health, Environment Protection, and Nutrition. The dataset provides a snapshot of the organizations working in the specified region and their focus areas.'; column_name:'Country'; ...   
350  [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/4W_BU_Kayanza.csv'; dataset_description='The dataset contains inform

(7, 4)


The model is correct here, the column names are typically things like 'Country'.

#### Scenario 3 - Where prediction add 'name' but labeled data doesn't have it

In [ ]:
breaks = breaks[~breaks.index.isin(scenario2.index)]
print(breaks.shape)

scenario3 = breaks[(breaks["predicted"].str.len() > breaks["actual"].str.len()) & (breaks["predicted"].str.contains("name"))]
display(scenario3)
print(scenario3.shape)

(133, 4)


prompt  \
1    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
51   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/victimas_hecho_dpto_2017_2021.xlsx'; dataset_description='The dataset "victimas_hecho_dpto_2017_2021.xlsx" contains information on victimizing events in Colombia from 2017 to 2021. It includes columns such as country, department code and name, year, type of victimizing event, gender, ethnicity, disability status, age range of victims, and number of individuals affected per occurrence. The data provides details on the location, characteristics, and impact of victimizing events, including information on the victims' demographics and vulnerabilities.'; column_name:'PAIS'; examples: ['COLOMBIA', 'COLOMBIA', ...   
88   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/victimas_explotacion_sexual_comercial.xlsx'; dataset_description='The dataset contains information on victims of commercial sexual exploitation, including details such as stage of legal process, department, municipality, charges, arrests, victim age group, country of birth, and total number of victims. The data includes entries for different stages of legal proceedings, with varying details for each victim such as age group and location. The dataset appears to have a mix of categorical and numerical data, with some missing values for certain variables like country of birth.'; column_name:'PAIS_NACIMIENTO...   
99   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/datos_brutos_personas_alcanzadas_vbg.xlsx'; dataset_description='The dataset contains information related to projects aimed at providing access to safe and confidential services for survivors of Gender Based Violence (GBV) in vulnerable populations during the COVID-19 pandemic. The data includes details such as project IDs, validation status, organization names, project descriptions, activity descriptions, resource types, estimated costs, start and end dates, status, country information, location details, number of people reached, demographics of beneficiaries, and indicators related to COVID-19, armed v...   
136  [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Base convalidaciones - HDX.xlsx'; dataset_description='The dataset t

(18, 4)


Looking at the above, apart from two real breaks, the rest seem to be correct. The model is adding 'name' for what are name columns.

#### The rest

In [ ]:
breaks = breaks[~breaks.index.isin(scenario3.index)]
print(breaks.shape)
display(breaks)
print(breaks.shape)

(115, 4)


prompt  \
6    [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators fo...   
12   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provi...   
13   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provi...   
16   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Human Development Indicators for Zimbabwe.csv'; dataset_description='The dataset contains Human Development Indicators for Zimbabwe, with information on indicators such as Adolescent Birth Rate (births per 1,000 women ages 15-19) for the years 1990 to 1998. The data includes columns for country code, country name, indicator ID, indicator name, index ID, index name, indicator value, and year. The values in the dataset show a trend of decreasing Adolescent Birth Rate over the specified years.'; column_name:'indicator_id'; examples: ['abr', 'abr', 'abr', 'abr', 'abr', 'abr', 'abr', 'abr', 'abr', 'abr', 'abr...   
20   [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Human Development Indicators for Somalia.csv'; dataset_description='

(115, 4)


The above is a bit of a mixed bag ...

- Some cases where the model is incorrect, typically where the model did not have enough context in the prompt. For example, rows above where actual is "#org" and predicted varies. The excel sheet has more surrounding context than provided to the model.

- There are also cases where the model seems reasonable (eg #beneficiary+type for column "Beneficiary type", #activity+type for column "Type of Activity").

- Administrative level disagreements, eg 'Province' column name where actual is #adm2+name, but the model predicts province as #adm1+name.

- Various other mismatches around tag order, differing granularity of tags and more

The takeaway for the exceptions is that in most cases, it's the human-labeled data that could be improved.

## Creating a better test set

Let's create a spreadsheet for human-review of the failed predictions.

In [ ]:
X_test = read_prompts_file(TEST_FILE)
X_test = pd.DataFrame(X_test)
X_test["actual"] = results["actual"]
X_test["predicted"] = results["predicted"]
X_test = X_test.drop(columns=["messages", "expected"])
X_test["match"] = X_test['predicted'] == X_test['actual']

display(X_test)
print(results.shape)
X_test.to_excel(f"{LOCAL_DATA_DIR}/hxl-metadata-prediction-test-modified.xlsx", index=False)


Found 437 prompts

Data providers ['dhs' 'undp-human-development-reports-office' 'fieldsdata' 'immap'
 'hera-humanitarian-emergency-response-africa' 'cimp' 'meers' 'rca'
 'global-shelter-cluster']

 tag
#adm1           58
#adm2           55
#country        54
#affected       43
#org            43
#date           30
#indicator      27
#region         24
#inneed         18
#sector         17
#meta           13
#loc             9
#adm3            8
#activity        8
#population      7
#beneficiary     6
#status          6
#reached         2
#targeted        2
#value           1
#event           1
#contact         1
#service         1
#delivery        1
#frequency       1
#output          1
Name: count, dtype: int64


Data description  \
0                      The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.   
1                      The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.   
2                      The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.   
3                      The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.   
4                      The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.   
..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        ...   
432  The dataset contains information related to humanitarian aid activities in Iraq in 2015, including details such as the date of ac

(437, 4)


Looking at the [human reviewed](https://docs.google.com/spreadsheets/d/1YwlubOYTVyipR26Drr8bqJae46dr7BuWgqmPbtEytRI/edit?usp=sharing) sheet, the breaks fall into a few categories ...


```
Fail reason,COUNTA of Fail reason
#org+url not set correctly,3
Actual is incorrect,12
Admin level mismatch,20
Admin1 vs. Location ambiguity,1
"Both prediction and actual have reasonable, but different, attributes",3
Not detecting #region,1
Not detecting region,20
"Prediction agrees, attribute order differs",1
Prediction has correct granular attributes,21
Prediction has extraneous incorrect attributes,4
Prediction incorrect,11
Prediction missing attribute granularity,30
Prompt lacks sufficient context of all columns,42
```

These categories are likely due to issues in the human labeled data and we should either update the 'actual' in the test set, or simply remove them.

There are two categories due to misidentifying region-related tags as well as admin levels. For these, we will improve the fine-tuning prompt.

Finally, many mismatching would seem to be where the data provided in the column alone is insifficient. It might be good to add ion a general description of the data.



## Just prompts, no fine-tuning



### GPT-4o

In [ ]:
def generate_hxl_standard_prompt(local_data_file):

  core_hashtags = pd.read_excel(local_data_file, sheet_name='Core hashtags')
  core_hashtags = core_hashtags.loc[core_hashtags["Release status"] == "Released"]
  core_hashtags = core_hashtags[["Hashtag", "Hashtag long description", "Sample HXL"]]

  core_attributes = pd.read_excel(local_data_file, sheet_name='Core attributes')
  core_attributes = core_attributes.loc[core_attributes["Status"] == "Released"]
  core_attributes = core_attributes[["Attribute", "Attribute long description", "Suggested hashtags (selected)"]]

  print(core_hashtags.shape)
  print(core_attributes.shape)

  core_hashtags = core_hashtags.to_dict(orient='records')
  core_attributes = core_attributes.to_dict(orient='records')

  hxl_prompt= f"""
  You are an AI assistant that predicts Humanitarian Markup Language (HXL) tags and attributes for columns of data where the HXL standard is defined as follows:

  CORE HASHTAGS:

  {json.dumps(core_hashtags,indent=4)}

  CORE ATTRIBUTES:

  {json.dumps(core_attributes, indent=4)}

  Key points:

  - ALWAYS predict hash tags
  - NEVER predict a tag which is not a valid core hashtag
  - NEVER start with a core hashtag, you must always start with a core hashtag
  - Always try and predict an attribute if possible

  You must return your result as a JSON record with the fields 'predicted' and 'reasoning', each is of type string.

  """

  print(len(hxl_prompt.split(" ")))
  print(hxl_prompt)
  return hxl_prompt

def call_gpt(prompt, system_prompt, model, temperature, top_p, max_tokens):
    response = client.chat.completions.create(
        model=model,
        messages= [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": prompt}
        ],
        max_tokens=2000,
        temperature=temperature,
        top_p=top_p,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None,
        stream=False,
        response_format={ "type": "json_object" }
    )

    result = response.choices[0].message.content
    result = result.replace("```json","").replace("```","")
    try:
        result = json.loads(result)
        result["predicted"] = result["predicted"].replace(" ","")
    except:
        print(result)
        result = None
    return result

def make_prompt_predictions(prompts, model, temperature=0.1, top_p=0.1, max_tokens=2000):

  results = []
  for index, p in prompts.iterrows():

      prompt = p["messages"][1]["content"]
      actual = p["messages"][2]["content"]

      result = call_gpt(prompt, hxl_prompt, model, temperature, top_p, max_tokens)

      if result is None:
          print("    !!!!! No LLM result")
          predicted = ""
          reasoning = ""
      else:
          predicted = result["predicted"]
          reasoning = result["reasoning"]

      results.append({
          "prompt": prompt,
          "actual": actual,
          "predicted": predicted,
          "reasoning": reasoning
      })

      #pprint.pp(prompt)
      #print(f'   Actual.   : {actual}')
      #print(f'   Predicted : {predicted}')
      #pprint.pp(f'   Reasoning : {reasoning}')
      #print()

  results = pd.DataFrame(results)

  print(f"\n\n===================== {model} Results =========================\n\n")
  output_prediction_metrics(results)
  print(f"\n\n=================================================================")

  results["match"] = results['predicted'] == results['actual']
  results.to_excel(f"{LOCAL_DATA_DIR}/hxl-metadata-prompting-only-prediction-{model}-results.xlsx", index=False)

  return results

hxl_prompt = generate_hxl_standard_prompt(local_data_file)


(44, 3)
(54, 3)
5240

  You are an AI assistant that predicts Humanitarian Markup Language (HXL) tags and attributes for columns of data where the HXL standard is defined as follows:

  CORE HASHTAGS:

  [
    {
        "Hashtag": "#access",
        "Hashtag long description": "Accessiblity and constraints on access to a market, distribution point, facility, etc.",
        "Sample HXL": "#access +type"
    },
    {
        "Hashtag": "#activity",
        "Hashtag long description": "A programme, project, or other activity. This hashtag applies to all levels; use the attributes +activity, +project, or +programme to distinguish different hierarchical levels.",
        "Sample HXL": "#activity +project"
    },
    {
        "Hashtag": "#adm1",
        "Hashtag long description": "Top-level subnational administrative area (e.g. a governorate in Syria).",
        "Sample HXL": "#adm1 +code"
    },
    {
        "Hashtag": "#adm2",
        "Hashtag long description": "Second-level subnationa

In [ ]:
for model in ["gpt-4o-mini","gpt-4o"]:
  print(f"Model: {model}")
  results = make_prompt_predictions(X_test, model, temperature=0.1, top_p=0.1, max_tokens=2000)

Model: gpt-4o-mini


===================== gpt-4o-mini Results =========================


LLM results for predicted, 437 predictions ...

Just HXL tags ...

Accuracy: 0.66
Precision: 0.74
Recall: 0.66
F1: 0.65

Tags and attributes with predicted ...

Accuracy: 0.39
Precision: 0.41
Recall: 0.39
F1: 0.36


Model: gpt-4o
    !!!!! No LLM result


===================== gpt-4o Results =========================


LLM results for predicted, 437 predictions ...

Just HXL tags ...

Accuracy: 0.82
Precision: 0.84
Recall: 0.82
F1: 0.82

Tags and attributes with predicted ...

Accuracy: 0.48
Precision: 0.52
Recall: 0.48
F1: 0.46




#### Prediction analysis

*How does prompting-only GPT-4o compare with GPT-4o-mini?*  

Looking at the above, we see that GPT-4o-mini prompting-only performs does not perform as well as GPT-4o prompting-only or GPT-4o fine-tuned, so doesn't seem to be a viable option for this task if prompting-only.

*How does prompting-only compare with the fine-tuned model?*  

Fine-tuning GPT-4o-mini yielded these results ...

```
Just HXL tags ...

Accuracy: 0.81
Precision: 0.84
Recall: 0.81
F1: 0.81

Tags and attributes with predicted ...

Accuracy: 0.62
Precision: 0.64
Recall: 0.62
F1: 0.6
```

GPT-4o prompting-only gave these results ...

```
Just HXL tags ...

Accuracy: 0.82
Precision: 0.84
Recall: 0.82
F1: 0.82

Tags and attributes with predicted ...

Accuracy: 0.48
Precision: 0.52
Recall: 0.48
F1: 0.46

```

Which gives comparible performance for tags-only, but poorer performance when considering tags and attributes. Let's take a close look at the latter ...

In [ ]:
X_test = read_prompts_file(TEST_FILE)
X_test = pd.DataFrame(X_test)

df = pd.read_excel(f"{LOCAL_DATA_DIR}/hxl-metadata-prompting-only-prediction-gpt-4o-results.xlsx")

# Concat X_test and results by columns
df = pd.concat([df,X_test], axis=1)

breaks = df[df["match"]==False]
print(breaks.shape)
display(breaks)


Found 437 prompts

Data providers ['dhs' 'undp-human-development-reports-office' 'fieldsdata' 'immap'
 'hera-humanitarian-emergency-response-africa' 'cimp' 'meers' 'rca'
 'global-shelter-cluster']

 tag
#adm1           58
#adm2           55
#country        54
#affected       43
#org            43
#date           30
#indicator      27
#region         24
#inneed         18
#sector         17
#meta           13
#loc             9
#adm3            8
#activity        8
#population      7
#beneficiary     6
#status          6
#reached         2
#targeted        2
#value           1
#event           1
#contact         1
#service         1
#delivery        1
#frequency       1
#output          1
Name: count, dtype: int64
(229, 18)


prompt  \
1    What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.'; column_name:'Location'; examples: ['Região Centro', 'Região Sul...   
6    What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe.csv'; dataset_description='The dataset contains information on various indicators related to health and demographics in Sao Tome and Principe, with data points such as total fertility rate and contraceptive use among married women across different regions. The data includes details like survey year, survey ID, indicator ID, indicator order, and precision. Each data point is associated with specific regions within the country and includes values, characteristic labels, and survey type. The dataset provides insights into key health and demographic indicators for analysis and comparison across different regions of Sao Tome and Principe.'; column_name:'ByVariableId'; examples: ['0', '0', '0', '0', '0',...   
12   What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provides a comprehensive overview of key health and demographic indicators for Sao Tome and Principe in 2008.'; column_name:'ByVariableId'; example...   
13   What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DHS Quickstats Data for Sao Tome and Principe1.csv'; dataset_description='The dataset contains various indicators related to demographic and health statistics for Sao Tome and Principe in 2008, sourced from the DHS Quickstats Data file. The indicators include total fertility rate, contraceptive use among married women, unmet need for family planning, median age at first marriage and first sexual intercourse for women, and infant mortality rates. Each data entry includes information such as the indicator name, value, precision, country code, survey details, and characteristic categories. The dataset provides a comprehensive overview of key health and demographic indicators for Sao Tome and Principe in 2008.'; column_name:'ByVariableLabel'; exam...   
16                                                                                                                                                  What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/Human Development Indicators for Zimbabwe.csv'; dataset_description

##### Scenario 1 - GPT-4o predicted the correct attribute, labelled data incorrectly was missing this

Looking at the results we see cases where GPT-4o predicted an additional attribute.

Let's take a closer look at these ...



In [ ]:
##### Scenario 1 - GPT-4o predicted the correct attribute, labelled data incorrectly was missing this



In [ ]:
df['actual_tag'] = results['actual'].apply(lambda x: x.split('+')[0])
df['predicted_tag'] = results['predicted'].apply(lambda x: x.split('+')[0])

# Find examples where tag is the same, but predicted has more attributes (ie more '+')
df2 = df[(df['actual_tag'] == df['predicted_tag']) & (df['predicted'].str.count('\+') > df['actual'].str.count('\+'))]
display(df2)

prompt  \
32                                                                                      What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/PiN_VBG_2023_hdx.xlsx'; dataset_description='The dataset from the file "PiN_VBG_2023_hdx.xlsx" contains information on different municipalities within the Amazonas region, including their respective DIVIPOLA codes, population in 2022, severity level based on the Joint Intersectoral Analysis Framework (JIAF) scale, and the number of people in need (PiN). The severity levels range from 2 to 4, with corresponding PiN values provided. The dataset provides a snapshot of the population in need across various municipalities in the Amazonas region for the year 2022.'; column_name:'Población 2022'; examples: [2111, 1059, 4005, 673, 1910, 1801, 10705, 4013, 3004, 784, 52003]   
48   What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/homicidios_mujeres_feminicidios_EneroMarzo_2022.csv'; dataset_description='The dataset contains information on homicides and feminicides of women in different departments and municipalities in Colombia from January to March 2022. It includes data such as the number of homicides and feminicides, total female population, and rates of homicides and feminicides per 100,000 women. The dataset also includes columns for department, municipality, Divipola code, and additional unnamed columns. The data provides insights into the prevalence of these crimes in various regions of the country during the specified time period.'; column_name:'Total Mujeres'; examples: ['27771', '4687', '14749', '27668', '154278', '11036', '1359799', '3970', '20628', '15062',...   
49   What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/homicidios_mujeres_feminicidios_EneroMarzo_2022.csv'; dataset_description='The dataset contains information on homicides and feminicides of women in different departments and municipalities in Colombia from January to March 2022. It includes data such as the number of homicides and feminicides, total female population, and rates of homicides and feminicides per 100,000 women. The dataset also includes columns for department, municipality, Divipola code, and additional unnamed columns. The data provides insights into the prevalence of these crimes in various regions of the country during the specified time period.'; column_name:'Tasa_homicidios'; examples: ['3.600878614', '21.33560913', '6.780120686', '3.614283649', '0.648180557', '9.061254078'...   
50   What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/homicidios_mujeres_feminicidios_EneroMarzo_2022.csv'; dataset_description='The dataset contains information on homicides and feminicides of women in different departments and municipalities in Colombia from January to March 2022. It includes data such as the number of homicides and feminicides, total female population, and rates of homicides and feminicides per 100,000 women. The dataset also includes columns for department, municipality, Divipola code, and additional unnamed columns. The data provides insights into the prevalence of these crimes in various regions of the country during the specified time period.'; column_name:'tasa_feminicidios'; examples: ['3.600878614', '0', '6.780120686', '0', '0.648180557', '0', '0.073540281', '0', '0', '...   
51                                         What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/victimas_hecho_dpto_2017_2021.xlsx'; dataset_description='The dataset "victimas_hecho_dpto_2017_2021.xlsx" contains information on victimizing events in Colombia from 2017 

129 out of the 229 predictions which disagreed with the labelled data fall into this category.

Let's print, including why the LLM did what it did ...

In [ ]:
for index, row in breaks.iterrows():
  print("\n")
  print(f"Actual", row["actual"])
  print(f"Predicted", row["predicted"])
  print(f"Column name: ", row["Text header"])
  print(f"Data excerpt: ", row["Data excerpt"])
  pprint.pp(f'\nReasoning: \n{row["reasoning"]}')




Actual #loc+name
Predicted #adm1+name
Column name:  Location
Data excerpt:  ['Região Centro', 'Região Sul', 'Região Norte', 'Região do Principe', 'Região Centro', 'Região Sul', 'Região Norte', 'Região do Principe', 'Região Centro', 'Região Sul', 'Região Norte']
('\n'
 'Reasoning: \n'
 "The column 'Location' contains names of regions within Sao Tome and "
 "Principe, such as 'Região Centro', 'Região Sul', 'Região Norte', and 'Região "
 "do Principe'. These are top-level subnational administrative areas, which "
 'corresponds to the HXL tag #adm1. The attribute +name is used because the '
 'data consists of human-readable names of these regions.')


Actual #indicator+code+label
Predicted #indicator+code
Column name:  ByVariableId
Data excerpt:  ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0']
('\n'
 'Reasoning: \n'
 "The column 'ByVariableId' appears to contain unique identifiers for "
 'different indicators related to health and demographics. The values in the '
 "column are nu

There are some genuine discrepancies, but the majority are cases like this ...

```
Actual #affected+infected
Predicted #affected+infected+num
Column name:  new_infected
Data excerpt:  ['1', '8', '3', '4', '2', '2', '3', '1', '1', '0', '2']
('\n'
 'Reasoning: \n'
 "The column 'new_infected' represents the number of new COVID-19 infections. "
 'The appropriate core hashtag is #affected, as it deals with the number of '
 'people affected by an emergency. The attribute +infected specifies that '
 'these are people infected with a disease, and +num indicates that the data '
 'consists of quantitative, numeric information.')
```

Where the predicted attributes are actually correct. In the example above '+num' is a valid attribute for a column providing the number of affected people.




##### Scenario 2 - Cases where GPT-4o predicted admin1 where the data was labelled 'region'

In [ ]:
# Find examples where tag in actual_tag is region and predicted is adm1
breaks = df[(df['actual_tag'] == '#region') & (df['predicted_tag'] == '#adm1')]

for index, row in breaks.iterrows():
  print("\n")
  print(f"Actual", row["actual"])
  print(f"Predicted", row["predicted"])
  print(f"Column name: ", row["Text header"])
  print(f"Data excerpt: ", row["Data excerpt"])
  pprint.pp(f'\nReasoning: \n{row["reasoning"]}')



Actual #region+name
Predicted #adm1+name
Column name:  Region
Data excerpt:  ['Muyinga', 'Muyinga', 'Muyinga', 'Muyinga', 'Cankuzo', 'Rumonge', 'Ruyigi', 'Ruyigi', 'Makamba', 'Makamba', 'Rutana']
('\n'
 'Reasoning: \n'
 "The column 'Region' contains names of top-level subnational administrative "
 'areas (e.g., Muyinga, Cankuzo, Makamba, etc.) in Burundi. According to the '
 'HXL standard, the appropriate hashtag for top-level subnational '
 'administrative areas is #adm1, and the attribute +name is used to indicate '
 'human-readable names.')


Actual #region+name
Predicted #adm1+name
Column name:  REGION
Data excerpt:  ['Abia', 'Adamawa', 'Akwa Ibom', 'Anambra', 'Bauchi', 'Bayelsa', 'Benue', 'Borno', 'Cross River', 'Delta', 'Ebonyi']
('\n'
 'Reasoning: \n'
 "The column 'REGION' contains names of top-level subnational administrative "
 'areas (e.g., states in Nigeria). According to the HXL standard, the '
 'appropriate hashtag for top-level subnational administrative areas is #adm1,

It seems GPT-4o is predicting administrative level 1 where the labelled data specifies region. However, looking at the above examples, they are correct. Even though they are technically regions, they are also the top administrative level for the country in question. Admin level 1 can vary by country, some it may be state, but for others, region.

So it would appear that GPT-4o is correct. Additionally, if we are trying to link together datasets geographically, it makes sense to use administrative levels consistents across countries, so in the case GPT-4o may even be producing a more useful result.



# Generate a Fine-Tuning Dataset

Let's use the LLM method to generate a fine-tuning dataset. We will enrich with tags from the HXL standard

## Find tags and attributes which are not in the training dataset

In [124]:
hxl_tags_attributes = []
for index, row in X_test.iterrows():
  for item in row["expected"].split("+"):
    if not item.startswith("#"):
      item = f"#{item}"
    hxl_tags_attributes.append(item)

hxl_tags_attributes = list(set(hxl_tags_attributes))
hxl_tags_attributes.sort()
print(hxl_tags_attributes)

newset = []
for item in APPROVED_HXL_SCHEMA:
  if item in hxl_tags_attributes:
    # Find rows in training set which have item in 'expected' column
    print(item)
    records = X_test[X_test["expected"].str.contains(item)]
    newset.append(records)
  if item not in hxl_tags_attributes:
    print(item)
    sys.exit()

['#activity', '#adm1', '#adm2', '#adm3', '#adults', '#affected', '#beneficiary', '#children', '#cluster', '#code', '#contact', '#country', '#date', '#delivery', '#elderly', '#end', '#event', '#f', '#frequency', '#id', '#impl', '#indicator', '#infected', '#inneed', '#killed', '#label', '#loc', '#m', '#meta', '#name', '#num', '#occurred', '#org', '#output', '#population', '#project', '#provided', '#reached', '#region', '#reporting', '#sector', '#service', '#source', '#start', '#status', '#targeted', '#text', '#type', '#url', '#value']
#access


SystemExit: 

/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py:3561: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [104]:
display(X_train)

messages  \
0     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx'; dataset_description='The dataset from the file "DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx" contains information on the total number of internally displaced persons (IDPs) and returnees in the Democratic Republic of Congo. The data includes the total number of IDP households, IDP individuals, male and female IDPs, and returnees. Specifically, there are 319,283 IDP households, 1,548,732 IDP individuals, with 646,805 males and 901,927 females, and 587,705 returnees.'; column_name:'Total IDP HH'; examples: [319283]'}, {'rol...   
1     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx'; dataset_description='The dataset from the file "DRC - Baseline Assessment - M23 Crisis 13 - February 2024.xlsx" contains information on the total number of internally displaced persons (IDPs) and returnees in the Democratic Republic of Congo. The data includes the total number of IDP households, IDP individuals, male and female IDPs, and returnees. Specifically, there are 319,283 IDP households, 1,548,732 IDP individuals, with 646,805 males and 901,927 females, and 587,705 returnees.'; column_name:'Total Returnees'; examples: [587705]'}, {'...   
2     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The dataset contains earthquake data for various administrative regions in Afghanistan, including country name, admin1 name, latitude, longitude, aggregation type, indicator name, and indicator value. The data includes maximum earthquake values recorded in different regions, with corresponding latitude and longitude coordinates. The dataset provides insights into the seismic activity in different administrative areas of Afghanistan.'; column_name:'country_name'; examples: ['Afghanistan', 'Afghanistan', 'Afghanistan', 'Afghanistan', 'Afghanistan', 'Af...   
3     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The dataset contains earthquake data for various administrative regions in Afghanistan, including country name, admin1 name, latitude, longitude, aggregation type, indicator name, and indicator value. The data includes maximum earthquake values recorded in different regions, with corresponding latitude and longitude coordinates. The dataset provides insights into the seismic activity in different administrative areas of Afghanistan.'; column_name:'latitude'; examples: ['34.5527', '34.9568', '34.9619', '34.3033', '34.0121', '34.2743', '34.7693', '35.4...   
4     [{'role': 'system', 'content': '
        You are an assistant that replies with HXL tags and attributes"
    '}, {'role': 'user', 'content': 'What are the HXL tags and attributes for a column with these details? resource_name='/content/drive/MyDrive/Colab/hxl-metadata-prediction/data/admin1-summaries-earthquake.csv'; dataset_description='The da